File: src/bumperbot_controllers/src/simple_controller.cpp
#include <bumperbot_controllers/simple_controller.hpp>
#include <Eigen/Geometry>
#include <tf2/LinearMath/Quaternion.hpp>

using std::placeholders::_1;

SimpleController::SimpleController(const std::string &name) : Node(name), 
    left_wheel_previous_pos_(0.0), right_wheel_previous_pos_(0.0),
    x_(0.0), y_(0.0), theta_(0.0)
{
    declare_parameter("wheel_radius", 0.033);
    declare_parameter("wheel_separation", 0.17);

    get_parameter("wheel_radius", wheel_radius_);
    get_parameter("wheel_separation", wheel_separation_);

    RCLCPP_INFO(this->get_logger(), "Wheel radius: %f", wheel_radius_);
    RCLCPP_INFO(this->get_logger(), "Wheel separation: %f", wheel_separation_);

    wheel_cmd_pub_ = this->create_publisher<std_msgs::msg::Float64MultiArray>("/simple_velocity_controller/commands", 10);

    velocity_sub_ = this->create_subscription<geometry_msgs::msg::TwistStamped>(
        "/bumperbot_controller/cmd_vel", 10, std::bind(&SimpleController::velocityCallback, this, std::placeholders::_1));

    joint_state_sub_ = this->create_subscription<sensor_msgs::msg::JointState>(
        "/joint_states", 10, std::bind(&SimpleController::jointStateCallback, this, std::placeholders::_1));
    
    odometry_pub_ = this->create_publisher<nav_msgs::msg::Odometry>(
        "/bumperbot_controller/odom", 10
    );

    previous_time_ = this->now();

    odom_msg_.header.frame_id = "odom";
    odom_msg_.child_frame_id = "bumperbot_footprint";
    odom_msg_.pose.pose.orientation.x = 0.0;
    odom_msg_.pose.pose.orientation.y = 0.0;
    odom_msg_.pose.pose.orientation.z = 0.0;
    odom_msg_.pose.pose.orientation.w = 1.0;

    // Initialize the speed conversion matrix
    speed_conversion_matrix_ << wheel_radius_ / 2.0,
        wheel_radius_ / 2.0,
        wheel_radius_ / wheel_separation_,
        -wheel_radius_ / wheel_separation_;

    transform_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);
    transform_stamped_.header.frame_id = "odom";
    transform_stamped_.child_frame_id = "base_footprint";

    RCLCPP_INFO_STREAM(this->get_logger(), "Conversion Matrix: \n" << speed_conversion_matrix_);
}

void SimpleController::velocityCallback(const geometry_msgs::msg::TwistStamped &msg)
{
    // Convert the velocity command to wheel speeds
    Eigen::Vector2d robot_speed ( msg.twist.linear.x, msg.twist.angular.z);
    Eigen::Vector2d wheel_speed = speed_conversion_matrix_.inverse() * robot_speed;        

    // Create a message to publish the wheel speeds
    std_msgs::msg::Float64MultiArray wheel_cmd_msg;
    wheel_cmd_msg.data.push_back(wheel_speed.coeff(1)); // Left wheel speed
    wheel_cmd_msg.data.push_back(wheel_speed.coeff(0)); // Right wheel speed

    // Publish the wheel speeds
    wheel_cmd_pub_->publish(wheel_cmd_msg);
}

void SimpleController::jointStateCallback(const sensor_msgs::msg::JointState &msg)
{
    double dp_left = msg.position.at(1) - left_wheel_previous_pos_;
    double dp_right = msg.position.at(0) - right_wheel_previous_pos_;

    rclcpp::Time msg_time = msg.header.stamp;
    rclcpp::Duration dt = msg_time - previous_time_;

    previous_time_ = msg_time;
    left_wheel_previous_pos_ = msg.position.at(1);
    right_wheel_previous_pos_ = msg.position.at(0);

    double fi_left = dp_left / dt.seconds();
    double fi_right = dp_right / dt.seconds();
    
    auto linear = (wheel_radius_ * (fi_left + fi_right)) / 2.0;
    auto angular = (wheel_radius_ * (fi_right - fi_left)) / wheel_separation_;

    double d_s = (wheel_radius_ * (dp_left + dp_right)) / 2.0;
    double d_theta = (wheel_radius_ * (dp_right - dp_left)) / wheel_separation_;

    theta_ += d_theta;
    x_ += d_s * cos(theta_);
    y_ += d_s * sin(theta_);

    // RCLCPP_INFO_STREAM(rclcpp::get_logger("SimpleController"), 
    // "\nLinear: " << linear 
    // << "\nAngular: " << angular
    // << "\nX: " << x_
    // << "\nY: " << y_
    // << "\ntheta: " << theta_);


    odom_msg_.header.stamp = get_clock()->now();
    tf2::Quaternion q;
    q.setRPY(0,0,theta_);
    odom_msg_.pose.pose.orientation.x = q.x();
    odom_msg_.pose.pose.orientation.y = q.y();
    odom_msg_.pose.pose.orientation.z = q.z();
    odom_msg_.pose.pose.orientation.w = q.w();
    odom_msg_.pose.pose.position.x = x_;
    odom_msg_.pose.pose.position.y = y_;
    odom_msg_.twist.twist.linear.x = linear;
    odom_msg_.twist.twist.angular.z = angular;

    odometry_pub_->publish(odom_msg_);

    transform_stamped_.transform.translation.x = x_;
    transform_stamped_.transform.translation.y = y_;
    transform_stamped_.transform.rotation.x = q.getX();
    transform_stamped_.transform.rotation.y = q.getY();
    transform_stamped_.transform.rotation.z = q.getZ();
    transform_stamped_.transform.rotation.w = q.getW();
    transform_stamped_.header.stamp = get_clock()->now();
    transform_broadcaster_->sendTransform(transform_stamped_);
}


int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleController>("simple_controller");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/bumperbot_controllers/bumperbot_controllers/simple_controller.py
#!/usr/bin/env python3
import rclpy
import numpy as np
import math

from rclpy.node import Node
from std_msgs.msg import Float64MultiArray
from geometry_msgs.msg import TwistStamped
from sensor_msgs.msg import JointState
from nav_msgs.msg import Odometry
from rclpy.time import Time
from rclpy.constants import S_TO_NS
from tf_transformations import quaternion_from_euler

class SimpleController(Node):
    def __init__(self):
        super().__init__("simple_controller")
        self.declare_parameter("wheel_radius", 0.033)
        self.declare_parameter("wheel_separation", 0.17)

        self.x_ = 0.0
        self.y_ = 0.0
        self.theta = 0.0

        self.odom = Odometry()
        self.odom.header.frame_id = "odom"
        self.odom.child_frame_id = "base_footprint"
        self.odom.pose.pose.orientation.x = 0.0
        self.odom.pose.pose.orientation.y = 0.0
        self.odom.pose.pose.orientation.z = 0.0
        self.odom.pose.pose.orientation.w = 1.0

        self.wheel_radius_ = (
            self.get_parameter("wheel_radius").get_parameter_value().double_value
        )
        self.wheel_separation_ = (
            self.get_parameter("wheel_separation").get_parameter_value().double_value
        )

        self.get_logger().info(
            f"Paramters:\n\t Wheel Radius: {self.wheel_radius_}\n\t Wheel Separation: {self.wheel_separation_}"
        )

        self.left_wheel_previous_pos_ = 0.0
        self.right_wheel_previous_pos_ = 0.0

        self.previous_time_ = self.get_clock().now()

        self.sub = self.create_subscription(
            TwistStamped, "bumperbot_controller/cmd_vel", self.velocityCallback, 10
        )

        self.joint_subscriber_ = self.create_subscription(
            JointState, "joint_states", self.jointCallback, 10
        )

        self.command_pub_ = self.create_publisher(
            Float64MultiArray, "simple_velocity_controller/commands", 10
        )

        self.odom_pub_ = self.create_publisher(
            Odometry, "bumperbot_controller/odom", 10
        )

        self.speed_conversion_ = np.array(
            [
                [self.wheel_radius_ / 2, self.wheel_radius_ / 2],
                [
                    self.wheel_radius_ / self.wheel_separation_,
                    -self.wheel_radius_ / self.wheel_separation_,
                ],
            ],
        )

        self.get_logger().info(
            f"""
            Conversion Matrix:\n\t
            {self.speed_conversion_}\n
            """
        )
        self.get_logger().info("Simple Controller Node has been started.")

    def velocityCallback(self, msg):
        robot_speed = np.array([[msg.twist.linear.x], [msg.twist.angular.z]])
        wheel_speed = np.matmul(np.linalg.inv(self.speed_conversion_), robot_speed)

        wheel_speed_msg = Float64MultiArray()
        wheel_speed_msg.data = [wheel_speed[1, 0], wheel_speed[0, 0]]
        self.command_pub_.publish(wheel_speed_msg)

    def jointCallback(self, msg):
        dp_left = msg.position[1] - self.left_wheel_previous_pos_
        dp_right = msg.position[0] - self.right_wheel_previous_pos_
        dt = Time.from_msg(msg.header.stamp) - self.previous_time_

        self.left_wheel_previous_pos_ = msg.position[1]
        self.right_wheel_previous_pos_ = msg.position[0]
        self.previous_time_ = Time.from_msg(msg.header.stamp)

        fi_left = dp_left / dt.nanoseconds / S_TO_NS
        fi_right = dp_right / dt.nanoseconds / S_TO_NS

        linear = self.wheel_radius_ * (fi_right + fi_left) / 2
        angular = self.wheel_radius_ * (fi_right - fi_left) / self.wheel_separation_

        d_s = self.wheel_radius_ * (dp_right + dp_left) / 2
        d_theta = self.wheel_radius_ * (dp_right - dp_left) / self.wheel_separation_

        self.theta += d_theta
        self.x_ += d_s * math.cos(self.theta)
        self.y_ += d_s * math.sin(self.theta)

        self.get_logger().info(
            f"""
        \nlinear: {linear}
        \nangular: {angular}
        \nx: {self.x_}
        \ny: {self.y_}
        \ntheta: {self.theta}
        """
        )

        q = quaternion_from_euler(0.0,0.0,self.theta)
        self.odom.pose.pose.orientation.x = q[0]
        self.odom.pose.pose.orientation.y = q[1]
        self.odom.pose.pose.orientation.z = q[2]
        self.odom.pose.pose.orientation.w = q[3]
        self.odom.pose.pose.position.x = self.x_
        self.odom.pose.pose.position.y = self.y_
        self.odom.header.stamp = self.get_clock().now().to_msg()
        self.odom.twist.twist.linear.x = linear
        self.odom.twist.twist.angular.z = angular

        self.odom_pub_.publish(self.odom)

def main(args=None):
    rclpy.init(args=args)
    simple_controller = SimpleController()
    rclpy.spin(simple_controller)
    simple_controller.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()
